src/driver.cpp

```diff
@@ -1,165 +1,166 @@
+            << "  -llvm-dump   print the llvm module\n"
+            << "  -cfg-dump    print the control flow graph\n";
+}
+
+[[noreturn]] void error(std::string_view msg) {
+  std::cerr << "error: " << msg << '\n';
+  std::exit(1);
+}
+
+struct CompilerOptions {
+  std::filesystem::path source;
+  std::filesystem::path output;
+  bool displayHelp = false;
+  bool astDump = false;
+  bool resDump = false;
+  bool llvmDump = false;
+  bool cfgDump = false;
+};
+
+CompilerOptions parseArguments(int argc, const char **argv) {
+  CompilerOptions options;
+
+  int idx = 1;
+  while (idx < argc) {
+    std::string_view arg = argv[idx];
+
+    if (arg[0] != '-') {
+      if (!options.source.empty())
+        error("unexpected argument '" + std::string(arg) + '\'');
+
+      options.source = arg;
+    } else {
+      if (arg == "-h")
+        options.displayHelp = true;
+      else if (arg == "-o")
+        options.output = ++idx >= argc ? "" : argv[idx];
+      else if (arg == "-ast-dump")
+        options.astDump = true;
+      else if (arg == "-res-dump")
+        options.resDump = true;
+      else if (arg == "-llvm-dump")
+        options.llvmDump = true;
+      else if (arg == "-cfg-dump")
+        options.cfgDump = true;
+      else
+        error("unexpected option '" + std::string(arg) + '\'');
+    }
+
+    ++idx;
+  }
+
+  return options;
+}
+} // namespace
+
+int main(int argc, const char **argv) {
+  CompilerOptions options = parseArguments(argc, argv);
+
+  if (options.displayHelp) {
+    displayHelp();
+    return 0;
+  }
+
+  if (options.source.empty())
+    error("no source file specified");
+
+  if (options.source.extension() != ".yl")
+    error("unexpected source file extension");
+
+  std::ifstream file(options.source);
+  if (!file)
+    error("failed to open '" + options.source.string() + '\'');
+
+  std::stringstream buffer;
+  buffer << file.rdbuf();
+  SourceFile sourceFile{options.source.c_str(), buffer.str()};
+
+  Lexer lexer(sourceFile);
+  Parser parser(lexer);
+  auto [ast, success] = parser.parseSourceFile();
+
+  if (options.astDump) {
+    for (auto &&decl : ast)
+      decl->dump();
+    return 0;
+  }
+
+  if (!success)
+    return 1;
+
+  Sema sema(std::move(ast));
+  auto resolvedTree = sema.resolveAST();
+
+  if (options.resDump) {
+    for (auto &&decl : resolvedTree)
+      decl->dump();
+    return 0;
+  }
+
+  if (options.cfgDump) {
+    for (auto &&decl : resolvedTree) {
+      const auto *fn = dynamic_cast<const ResolvedFunctionDecl *>(decl.get());
+      if (!fn)
+        continue;
+
+      std::cerr << fn->identifier << ':' << '\n';
+      CFGBuilder().build(*fn).dump();
+    }
+    return 0;
+  }
+
+  if (resolvedTree.empty())
+    return 1;
+
+  Codegen codegen(std::move(resolvedTree), options.source.c_str());
+  llvm::Module *llvmIR = codegen.generateIR();
+
+  if (options.llvmDump) {
+    llvmIR->print(llvm::dbgs(), nullptr,
+      /*ShouldPreserveUseListOrder=*/false, /*IsForDebug=*/true);
```

Comment:
I think we should keep printing simple like this.

```suggestion
    llvmIR->print(llvm::errs(), nullptr);
```

The only noticeable change we get by dropping `IsForDebug` is that parameters in declarations are no longer named, but since in our case, it only affects external functions, I think it's fine.

```diff
-declare i32 @printf(i8* %0, ...)
+declare i32 @printf(i8*, ...)
```

Reply:
I did that at first but it breaks a test. I guess the test can be updated then.

Reply:
Yes, we can update. It broke for what I mentioned in my previous comment.